tests/framework/prototype.py

```diff
@@ -0,0 +1,208 @@
+            if int_step >= len(curr) or int_step < 0:
+                return None
+            curr = curr[int(step)]
+        elif step in curr:
+            curr = curr[step]
+        else:
+            return None
+    return curr
+
+def yaml_to_file(cr_yaml, fn):
+    f = open(fn, 'w')
+    f.write(yaml.dump(cr_yaml))
+    f.close()
+'''
+Functions for cluster preparation. Typical workflow:
+  Delete KinD cluster -> Create KinD cluster -> Install CRD -> Download Images (from DockerHub) ->
+  Load images into KinD cluster -> Install KubeRay operator
+'''
+def delete_kind_cluster():
+    os.system("kind delete cluster")
+
+def create_kind_cluster():
+    os.system("kind create cluster")
+    os.system("kubectl wait --for=condition=ready pod -n kube-system --all --timeout=900s")
+
+def install_crd():
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl create -k \"github.com/ray-project/kuberay/manifests/cluster-scope-resources?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def download_images(images):
+    client = docker.from_env()
+    for image in images:
+        client.images.pull(image)
+    client.close()
+
+def kind_load_images(images):
+    for image in images:
+        os.system('kind load docker-image {}'.format(image))
+
+def install_kuberay_operator():
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl apply -k \"github.com/ray-project/kuberay/manifests/base?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+    time.sleep(60)
+
+'''
+Configuration Test Framework Abstractions: (1) DeltaSet (2) Mutator (3) Rule (4) RuleSet (5) CREvent
+'''
+
+# DeltaSet: Use `path` to specify the field that wants to mutate, and `candidates` is a list of candidate
+#           value for the field.
+# Example:  DeltaSet("spec.headGroupSpec.template.spec.containers.0.name", ['ray-head-1', 'ray-head-2', 'ray-head-3'])
+class DeltaSet:
+    def __init__(self, path, candidates):
+        self.path = path
+        self.candidates = candidates
+
+    def iterate(self):
+        for candidate in self.candidates:
+            yield candidate
+
+    # return value: (1) steps (2) None => this path does not exist
+    def find_path(self, cr) -> Optional[list[str]]:
+        steps = self.path.split('.')
+        return steps
+
+# Mutator: Mutator will start to mutate from `baseCR`. `deltaSets` is a list of DeltaSets, and each DeltaSet
+#          specifies a field that wants to mutate with multiple candidate values.
+# Example: "SimpleMutator"
+class Mutator:
+    def __init__(self, baseCR, deltaSets: list[DeltaSet]):
+        self.baseCR = baseCR
+        self.deltaSets = deltaSets
+    # You need to define your mutate() function by inheriting `Mutator`. It should return a new CR.
+    def mutate(self):
+        pass
+    # Apply delta to the base (custom resource).
+    def apply_delta(self, base, delta, steps):
+        root = copy.deepcopy(base)
+        curr = search_path(root, steps[:-1])
```

Comment:
Use json patch to apply json patches.

Any series of changes can be achieved by applying a series of json patches to a base config.

Reply:
Added `jsonpatch` support in https://github.com/ray-project/kuberay/pull/605/commits/6e8b9a9ac8170d496c8531eb30979759dfbaaba5.